### PR TITLE
Added Bundle Assets - 

### PR DESCRIPTION
**Pull request by vkalpias**
Bundle Assets contain the files of multiple other assets in tar format. 

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: build/dependencies.txt [hunk 1]
```diff
@@ -199,6 +201,8 @@
 ../src/framework/utils/entity-reference.js
 ../src/framework/utils/sorted-loop-array.js
 ../src/resources/loader.js
+../src/resources/bundle.js
+../src/resources/untar.js
```
Comment (vkalpias) [inline build/dependencies.txt]: This library is not currently included.

Comment (guycalledfrank) [reply build/dependencies.txt]: Grouping multiple assets into large blobs is definitely a good idea. I wonder if you can just reuse existing browser GZIP compression and simplify the task into joining/splitting data without any fancy logic? E.g. you only need a file table with byte offsets + large chunk of glued data.

**Comment by guycalledfrank** [comment]
Grouping multiple assets into large blobs is definitely a good idea. I wonder if you can just reuse existing browser GZIP compression and simplify the task into joining/splitting data without any fancy logic? E.g. you only need a file table with byte offsets + large chunk of glued data.

File: src/resources/loader.js [hunk 2]
```diff
@@ -73,24 +75,63 @@ Object.assign(pc, function () {
...
+                        return;
                     }
-                    delete this._requests[key];
-                }.bind(this), asset);
+
+                    this._app.bundles.loadUrl(normalizedUrl, function (err, fileUrlFromBundle) {
+                        handleLoad(err, { load: fileUrlFromBundle, original: url });
+                    });
+                } else {
+                    handleLoad(null, { load: url, original: url });
```
Comment (daredevildave) [inline src/resources/loader.js]: Maybe we should define an actual class for this UrlObject? It's a bit confusing.

File: src/asset/asset.js [hunk 3]
```diff
@@ -147,17 +156,33 @@ Object.assign(pc, function () {
...
+                    var variant = VARIANT_DEFAULT_PRIORITY[i];
+                    // if the device supports the variant
+                    if (! device[VARIANT_SUPPORT[variant]]) continue;
+
+                    // if the variant exists in the asset then just return it
+                    if (this.file.variants[variant]) {
+                        return this.file.variants[variant];
+                    }
+
+                    // if the variant does not exist but we the asset is in a bundle
```
Comment (daredevildave) [inline src/asset/asset.js]: we *know* the asset is in a bundle

File: src/asset/asset.js [hunk 4]
```diff
@@ -147,17 +156,33 @@ Object.assign(pc, function () {
...
+                    // and the bundle contain assets with this variant then return the default
+                    // file for the asset
+                    if (app._enableBundles) {
```
Comment (daredevildave) [inline src/asset/asset.js]: This should be a public property if it's accessed outside the application

File: src/bundles/bundle-registry.js [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+        _getAssetFileUrls: function (asset) {
+            var url = asset.getFileUrl();
+            if (! url) return null;
+
+            url = this._normalizeUrl(url);
+            var urls = [url];
+
+            // a font might have additional files
+            // so add them in the list
+            if (asset.type === 'font') {
```
[the remark this answers is not in the document]
Comment (vkalpias) [reply src/bundles/bundle-registry.js]: I think eventually the possible URLs for assets should be in the asset. I could add a method on the asset like getFileUrls or something but I wanted to leave this until we actually do something about the font asset.

File: src/bundles/bundle-registry.js [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+            // to create its resource
+            if (! bundleAsset.resource) {
+                this._onBundleError('Bundle ' + bundleAsset.id + ' failed to load', bundleAsset);
+                return;
+            }
+
+            // on next tick resolve the pending asset requests
+            // don't do it on the same tick because that ties the loading
+            // of the bundle to the loading of all the assets
+            requestAnimationFrame(function () {
```
[the remark this answers is not in the document]
Comment (vkalpias) [reply src/bundles/bundle-registry.js]: Added a check to address this.

File: src/asset/asset.js (hunk 4 above)
Comment (vkalpias) [reply src/asset/asset.js]: Changed it to `enableBundles`

**Pull request merged**